File: baekjoon/mergeIntervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) ==0:
            return intervals
        
        #sort intervals
        intervals.sort(key=lambda x:x[0])
        
        
        start = intervals[0][0]
        end = intervals[0][1]
        
        
        
        ans = []
        for i in range(1,len(intervals)):
            if end < intervals[i][0]:
                ans.append([start, end])
                start = intervals[i][0]
                end = intervals[i][1]
            else:
                if intervals[i][1] < end:
                    continue
                end = intervals[i][1]
        ans.append([start,end])
        
        return ans

File: aexpt/ltcode/mergeIntervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    intervals.sort(
        (a,b) => a[0] - b[0]
    )
    
    const solution = []
    
    let curLow = -Infinity;
    let curMax = -Infinity;
    let curIndex = 0;
    let localMax = -Infinity;
    if(intervals.length === 0){
        return []
    }
    while(curIndex < intervals.length){
        if(curLow === -Infinity){
            curLow = intervals[curIndex][0]
            curMax = intervals[curIndex][1]
        }else{
            if(intervals[curIndex][0] <= curMax){
                curMax = Math.max(intervals[curIndex][1], curMax)
            }else{
                solution.push([curLow, curMax]);
                curLow = intervals[curIndex][0]
                curMax = intervals[curIndex][1]
            }
        }
        curIndex += 1
    }
    
    solution.push([curLow, curMax])
    return solution
};
